### ISSUE TITLE
LNCC loss instability for large kernel size

### ISSUE THREAD

**Issue by wyli** [Project-MONAI/MONAI#3803]
### Discussed in https://github.com/Project-MONAI/MONAI/discussions/3463

<div type='discussions-op-text'>

<sup>Originally posted by **ebrahimebrahim** December  9, 2021</sup>
I experienced some numerically unstable training while using LNCC loss to train a registration network. Exploding gradients, clipping them not helping, training loss increasing even with a low learning rate, etc.

This is in the context of synthetic data, where the images contained large homogeneous blobs, and so the variances used to compute LNCC were probably quite low.

I see that [small variance has been a problem in the past](https://github.com/Project-MONAI/MONAI/issues/1868) and that [it has been addressed](https://github.com/Project-MONAI/MONAI/pull/1915). However, I seem to still have the issue when I set the `kernel_size` argument for LNCC loss to be large. Setting it back to a small value like `3` fixes the problem completely. Setting it to a large value (say, half the image size) leads to problems.

I think that giving a large `kernel_size` to LNCC is a natural thing to do, because NCC (normalized cross correlation, without the "local") is the case of having `kernel_size` equal image size, if I'm understanding correctly.

_Is there any reason to expect that large kernel size would worsen the instability problem?_ From [the code](https://github.com/Project-MONAI/MONAI/blob/dev/monai/losses/image_dissimilarity.py#L50), it's not clear to me why the LNCC kernel size would matter. Thoughts, anyone?</div>

**Comment by wyli** [comment]
perhaps with large kernels, there's a problem with the zero padding here https://github.com/Project-MONAI/MONAI/blob/817c1cf1648417ed2688d7f9f8c01f0cb3eecbcf/monai/networks/layers/simplelayers.py#L208 @ebrahimebrahim

**Comment by ebrahimebrahim** [comment]
Copying this from the discussion: I have a notebook demonstrating the issue: https://github.com/ebrahimebrahim/deep-atlas/blob/main/lncc_with_large_kernel.ipynb



> perhaps with large kernels, there's a problem with the zero padding here?

:thinking:

**Comment by wyli** [comment]
thanks for the detailed demo, I did a quick change of the mode `zeros` to `circular` in `separable_filtering` it seems the loss becomes more reasonable
```diff
- def separable_filtering(x: torch.Tensor, kernels: List[torch.Tensor], mode: str = "zeros") -> torch.Tensor:
+ def separable_filtering(x: torch.Tensor, kernels: List[torch.Tensor], mode: str = "circular") -> torch.Tensor:
```

**Comment by Constantin-Jehn** [comment]
I would be interested in what's actually the stride when moving accross the image.
And couldn't it be interesting passing it as parameter to the loss?

**Comment by Constantin-Jehn** [comment]
I had similar problems for "big" kernel sizes.

I could overcome them by using the implementation from voxelmorph: https://github.com/voxelmorph/voxelmorph/blob/dev/voxelmorph/torch/losses.py

So maybe someone could compare to that code.

**Comment by wyli** [comment]
thanks for the discussion here, it's an issue handling the zero variance... I've included a patch in this PR to fix it https://github.com/Project-MONAI/MONAI/pull/5473.

@ebrahimebrahim I run your scripts and now the large kernels work fine. it's also quite efficient...

![Screenshot 2022-11-04 at 17 42 59](https://user-images.githubusercontent.com/831580/200041243-c203af4d-585b-43b7-b8f4-0f78de22b5e1.png)
![lncc](https://user-images.githubusercontent.com/831580/200041333-4bdcfcdd-3c4e-47f7-8824-11eef337600b.png)
